### ISSUE TITLE
Classes do not form lexical closures

### ISSUE THREAD

**Issue by cygx** [rakudo/rakudo#2690]
Example code:

    sub foo($a) {
        my class Foo {
            method a { $a }
        }
    };

    my $f1 = foo(1);
    my $f2 = foo(2);
    my $f3 = foo(3);

    say $f1.a, $f2.a, $f3.a  # yields:   333
                             # expected: 123

This might be 'broken' by design and not a bug as such. If so, is it possible to detect such cases and emit a warning?

    $ perl6 -v
    This is Rakudo version 2018.12-293-gd904b7048 built on MoarVM version 2018.12-94-g2eab275df
    implementing Perl 6.d.

**Comment by jnthn** [comment]
> This might be 'broken' by design and not a bug as such.

Yes, this is by design: classes are put together at compile-time, and composed at the closing `}` of their declaration. By contrast, taking of closures is a runtime thing. I don't think it's broken, just a place where there's more than one way things could have been done, and some folks might guess the wrong way if they don't know.

> If so, is it possible to detect such cases and emit a warning?

I think so, and I think it could be helpful if done right (we'd want to do it at a language version boundary, though, so it's a 6.e thing). The question is defining precisely what cases should warn. For example, this:

```
my $setting = %*ENV<MY_SETTING>;
class Foo {
    method m() {
        $setting ?? 1 !! 2
    }
}
```

Is something we'd probably consider annoying to get a warning about since it's harmless, as is:

```
class Foo {
    my $setting = %*ENV<MY_SETTING>;
    method m() {
        $setting ?? 1 !! 2
    }
}
```

Which is the same, but with a narrower scope. Nested classes are also not inherently problematic in this sense. So, we'd need to:

1. Formulate rules for when it should warn that will not cause too many false positives, but catch a decent number of the cases that are likely to be misunderstandings.
2. Implement it, but only warn then we're in `6.e` or above.

**Comment by vrurg** [comment]
I'd comment here, rather than in #2882 because this topic is more informative.

To my view, a warning should be issued if:

1. A symbol is lexically scoped.
1. Doesn't belong to the current package (class, role).
1. Isn't defined in a compunit.

As to the last item I'm not totally sure because something like the following code should be pretty much ok:

```
unit module Mine;

class Foo {
    my $foo;

    method fubar {
        my class Bar {
             method bar { $foo }
        }
    }
}
```

Perhaps it's a topic for problem solving?
